src/contracts/MsVault.h

```diff
@@ -0,0 +1,708 @@
+    _
+
+    // Procedures and functions
+    PUBLIC_PROCEDURE_WITH_LOCALS(registerVault)
+        if (qpi.invocationReward() < REGISTERING_FEE)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            return;
+        }
+
+        locals.ownerCount = 0;
+        for (locals.i = 0; locals.i < MAX_OWNERS; locals.i++)
+        {
+            if (input.owners.get(locals.i) != NULL_ID)
+                locals.ownerCount++;
+        }
+
+        if (locals.ownerCount <= 1)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            return;
+        }
+
+        // Find empty slot
+        locals.slotIndex = -1;
+        for (locals.i = 0; locals.i < MAX_VAULTS; locals.i++)
+        {
+            Vault tempVault = state.vaults.get(locals.i);
+            if (!tempVault.isActive && tempVault.numberOfOwners == 0 && tempVault.balance == 0)
+            {
+                locals.slotIndex = (sint32)locals.i;
+                break;
+            }
+        }
+
+        if (locals.slotIndex == -1)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            return;
+        }
+
+        locals.newVault.vaultType = input.vaultType;
+        locals.newVault.vaultName = input.vaultName;
+        locals.newVault.numberOfOwners = locals.ownerCount;
+        locals.newVault.balance = 0;
+        locals.newVault.isActive = true;
+
+        locals.rr_in.vault = locals.newVault;
+        resetReleaseRequests(qpi, state, locals.rr_in, locals.rr_out, locals.rr_locals);
+        locals.newVault = locals.rr_out.vault;
+
+        for (locals.i = 0; locals.i < locals.ownerCount; locals.i++)
+        {
+            locals.newVault.owners.set(locals.i, input.owners.get(locals.i));
+        }
+
+        state.vaults.set((uint64)locals.slotIndex, locals.newVault);
+
+        //qpi.burn(REGISTERING_FEE);
+        if (qpi.invocationReward() > REGISTERING_FEE)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward() - REGISTERING_FEE);
+        }
+
+        state.numberOfActiveVaults++;
+        state.totalRevenue += REGISTERING_FEE;
+    _
+
+    PUBLIC_PROCEDURE_WITH_LOCALS(deposit)
+        if (input.vaultID >= MAX_VAULTS)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            return;
+        }
+
+        locals.vault = state.vaults.get(input.vaultID);
+        if (!locals.vault.isActive)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            return;
+        }
+
+        locals.vault.balance += qpi.invocationReward();
+        state.vaults.set(input.vaultID, locals.vault);
+    _
+
+    PUBLIC_PROCEDURE_WITH_LOCALS(releaseTo)
+        if (qpi.invocationReward() > RELEASE_FEE)
+        {
+            qpi.transfer(qpi.invocator(), qpi.invocationReward() - RELEASE_FEE);
+        }
+        state.totalRevenue += RELEASE_FEE;
+
+        MSVaultLogger loggerMsg;
```

Comment:
Creating objects on the regular function call stack is not permitted. Use locals.loggerMsg.